compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs

```diff
@@ -490,17 +490,29 @@ impl<'a, 'tcx> ConstToPat<'a, 'tcx> {
                 // convert the dereferenced constant to a pattern that is the sub-pattern of the
                 // deref pattern.
                 _ => {
-                    let old = self.behind_reference.replace(true);
-                    // In case there are structural-match violations somewhere in this subpattern,
-                    // we fall back to a const pattern. If we do not do this, we may end up with
-                    // a !structural-match constant that is not of reference type, which makes it
-                    // very hard to invoke `PartialEq::eq` on it as a fallback.
-                    let val = match self.recur(tcx.deref_const(self.param_env.and(cv)), false) {
-                        Ok(subpattern) => PatKind::Deref { subpattern },
-                        Err(_) => PatKind::Constant { value: cv },
-                    };
-                    self.behind_reference.set(old);
-                    val
+                    if !pointee_ty.is_sized(tcx.at(span), param_env) {
```

Comment:
that certainly avoids the ICE, but we could also make such patterns legal. I think we should. @rust-lang/wg-const-eval what do you think about these?

Reply:
This always does `span_err` or `delay_span_bug`...  how does this make anything legal?

Reply:
I said "we *could* ...", I'm fine doing this PR, it changes and ICE to an error, but why should the code in the new test not be permitted? It seems perfectly alright to me

Reply:
I'm not familiar enough with pattern matching code to comment on this.

But I'd prefer we port that code to valtrees first, then I will be a lot more confident that what that code does makes sense in my mental model of rustc. :)

Reply:
@oli-obk What would be a good way to direct attention to this? I would find it incredibly useful to be able to use newtypes around `str` as const patterns.

Reply:
Valtrees are being worked on right now, once we have them, this becomes trivial to do and we can ask T-lang whether they think this should br supported, too